### PR TITLE
Plane: Allow a Lua script to return to a previous mission track

### PR DESCRIPTION
**Pull request by timtuxworth**
As discussed in #27639 on the dev call on 29 Aug 2024, added a way to get the previous WP location prior to switching out of AUTO mode and then restore it again after (in my case) doing some terrain object avoidance. This allows a plane to resume cross tracking mid mission which is required for magnetic field measurements in wilderness areas.

I changed the names proposed by @tridge in the call because:
1. getting the previous WP could conceivably be used for other purposes not just cross tracking, so I'm proposing vehicle:get_previous_location() - REMOVED based on the dev call on 12 Aug 2024
2. Restoring the previous location for crosstracking also requires setting the internal crosstrack boolean to true, so proposing to call this vehicle:set_crosstrack_start() which both sets the prev_WP_loc and also sets auto_state.crosstrack = true to resume cross tracking for the current mission item.

The sample script includes a LocationTracker() class which can be used to save previous target locations that can be passed to set_crosstrack_start()

Includes a sample lua script which requires you to be running a mission. A switch will save the previous location (high) and switch to loiter, then resume the mission and restore crosstracking (low). This screen shot shows a loiter being triggered, then switching back to crosstracking and resuming the path by crosstracking to the mission track rather than just heading direct for the next WP (WP 10).

Screenshot 2024-08-05 100641

### CODE REVIEW

File: ArduPlane/ArduPlane.cpp
Comment (IamPete1) [inline ArduPlane/ArduPlane.cpp:948]: This certainly wont do anything in some of these modes. Maybe just auto and the two RTLs?

Comment (timtuxworth) [reply ArduPlane/ArduPlane.cpp:948]: I'm going to say I'd want it in Guided.

Comment (timtuxworth) [reply ArduPlane/ArduPlane.cpp]: Fixed

**Comment by timtuxworth** [comment]
Tested on SITL and by Overhead Intelligence on a MFE Fighter VTOL

File: ArduPlane/ArduPlane.cpp
Comment (IamPete1) [inline ArduPlane/ArduPlane.cpp:926]: I think this should be the same modes as allow setting it. The problem is that if the mode does not use that location it really could be anywhere so you might end up navigating to somewhere new rather than somewhere you have already been.

File: ArduPlane/ArduPlane.cpp
Comment (IamPete1) [inline ArduPlane/ArduPlane.cpp:912]:
```suggestion
// allow script to find access prev_WP_loc
```

File: libraries/AP_Scripting/examples/crosstrack_restore.lua
Comment (IamPete1) [inline libraries/AP_Scripting/examples/crosstrack_restore.lua:36]: If nil you could use the current location.

File: ArduPlane/ArduPlane.cpp
Comment (IamPete1) [reply ArduPlane/ArduPlane.cpp:926]: TBH, I'm not really a fan of this method at all. We have a binding for current destination. You could easly use that in lua for the same result. 

If we were to add a binding for the `same_loc_as` location method:
```
local previous_location

local function update()

    local target = vehicle:get_target_location()
    if (target and (not previous_location or not previous_location:same_loc_as(target)) then
        previous_location = target
    end

   ....
end
```
If we dont add a binding for `same_loc_as` then you can add a little helper:
```
local function same_loc_as(A, B)
    if (A:lat() ~= B:lat()) or (A:lon() ~= B:lon()) then
        return false
    end
    return (A:alt() == B:alt()) and (A:get_alt_frame() == B:get_alt_frame())
end
```

File: libraries/AP_Scripting/docs/docs.lua
Comment (IamPete1) [inline libraries/AP_Scripting/docs/docs.lua:2505]:
```suggestion
---@param new_start_location Location_ud
```

File: ArduPlane/ArduPlane.cpp
Comment (timtuxworth) [reply ArduPlane/ArduPlane.cpp:926]: > I think this should be the same modes as allow setting it. The problem is that if the mode does not use that location it really could be anywhere so you might end up navigating to somewhere new rather than somewhere you have already been.

prev_WP_loc is set in Plane::update_GPS_10Hz as well as a number of other places like do_RTL(), do_takeoff(), verify_nav_wp() set_next_wp(), MAV_CMD_GUIDED_CHANGE_HEADING, ModeCruise::navigate()

My suggestion is to allow it to be used "as required" in Lua regardless of where its set. Why restrict it?

File: libraries/AP_Scripting/examples/crosstrack_restore.lua
Comment (timtuxworth) [reply libraries/AP_Scripting/examples/crosstrack_restore.lua:36]: Pretty sure that happens anyway, so I'm not sure why I would need/want to do this here.

File: ArduPlane/ArduPlane.cpp
Comment (timtuxworth) [reply ArduPlane/ArduPlane.cpp:926]: > TBH, I'm not really a fan of this method at all. We have a binding for current destination. You could easly use that in lua for the same result.
> 
> If we were to add a binding for the `same_loc_as` location method:
> 
> ```
> local previous_location
> 
> local function update()
> 
>     local target = vehicle:get_target_location()
>     if (target and (not previous_location or not previous_location:same_loc_as(target)) then
>         previous_location = target
>     end
> 
>    ....
> end
> ```
> 
> If we dont add a binding for `same_loc_as` then you can add a little helper:
> 
> ```
> local function same_loc_as(A, B)
>     if (A:lat() ~= B:lat()) or (A:lon() ~= B:lon()) then
>         return false
>     end
>     return (A:alt() == B:alt()) and (A:get_alt_frame() == B:get_alt_frame())
> end
> ```

I don't understand this

Comment (timtuxworth) [reply ArduPlane/ArduPlane.cpp:926]: > TBH, I'm not really a fan of this method at all. We have a binding for current destination. You could easly use that in lua for the same result.

As far as I can see prev_WP_loc may never have been a previous "target_location" based on all the places it is set in the code.

prev_WP_loc is set in Plane::update_GPS_10Hz as well as a number of other places like do_RTL(), do_takeoff(), verify_nav_wp() set_next_wp(), MAV_CMD_GUIDED_CHANGE_HEADING, ModeCruise::navigate() ...

File: libraries/AP_Scripting/examples/crosstrack_restore.lua
Comment (timtuxworth) [reply libraries/AP_Scripting/examples/crosstrack_restore.lua]: Fixed

File: libraries/AP_Scripting/examples/crosstrack_restore.lua
Comment (timtuxworth) [reply libraries/AP_Scripting/examples/crosstrack_restore.lua:2]: Fixed

File: libraries/AP_Scripting/examples/crosstrack_restore.lua
Comment (IamPete1) [reply libraries/AP_Scripting/examples/crosstrack_restore.lua:37]: In any of the modes form `get_previous_location` returns false then `previous_location` will be nil here.

Comment (timtuxworth) [reply libraries/AP_Scripting/examples/crosstrack_restore.lua:37]: Which is why line 31 tests for nil before using it. What would you do?

File: libraries/AP_Scripting/examples/crosstrack_restore.lua
Comment (IamPete1) [reply libraries/AP_Scripting/examples/crosstrack_restore.lua:2]: Sorry, I should have been more clear. 

I meant more that the script should not be used except on plane. It will run fine on other vehicles. As your comment points warns the functionality won't work. But it will change modes just fine. Except it won't be loiter and auto on other vehicles. It will be whatever modes happen have the same number.

You could check in the script and error out if not plane. I think this should work:
```
assert(FWVersion:type() == 3, "Must be plane")
```

File: libraries/AP_Scripting/examples/crosstrack_restore.lua
Comment (IamPete1) [reply libraries/AP_Scripting/examples/crosstrack_restore.lua:37]: It would just make the cross track from where if you flip the switch if your in a mode that does not support `get_previous_location`. 

I guess otherwise you could only accept the interrupt if in a mode that supports it. Here you always switch to loiter with the switch no matter if you have a valid cross track or not.

File: libraries/AP_Scripting/examples/crosstrack_restore.lua
Comment (timtuxworth) [reply libraries/AP_Scripting/examples/crosstrack_restore.lua:2]: This will throw and error but then the scheduler will try to run the script again. I used an if statement instead which only runs update() if the test passes.

File: ArduPlane/ArduPlane.cpp
Comment (tridge) [inline ArduPlane/ArduPlane.cpp:938]: i think we should remove the switch and unconditionally set prev_WP_loc and auto_state.crosstrack, and leave it up to script to determine when it is valid

File: ArduPlane/ArduPlane.cpp
Comment (tridge) [inline ArduPlane/ArduPlane.cpp:913]: I think Pete is right that this functionality can be done in the script without a binding

**Review by tridge — CHANGES_REQUESTED** [review]

File: ArduPlane/ArduPlane.cpp
Comment (timtuxworth) [reply ArduPlane/ArduPlane.cpp:926]: > TBH, I'm not really a fan of this method at all. We have a binding for current destination. You could easly use that in lua for the same result.
> 
> If we were to add a binding for the `same_loc_as` location method:
> 
> ```
> local previous_location
> 
> local function update()
> 
>     local target = vehicle:get_target_location()
>     if (target and (not previous_location or not previous_location:same_loc_as(target)) then
>         previous_location = target
>     end
> 
>    ....
> end
> ```
> 
> If we dont add a binding for `same_loc_as` then you can add a little helper:
> 
> ```
> local function same_loc_as(A, B)
>     if (A:lat() ~= B:lat()) or (A:lon() ~= B:lon()) then
>         return false
>     end
>     return (A:alt() == B:alt()) and (A:get_alt_frame() == B:get_alt_frame())
> end
> ```

I've thought about this a lot and I think the idea of duplicating code that already exists in c++ in Lua is a bad idea. In principle - if the functionality exists in c++ the correct solution is to create a binding.

Comment (timtuxworth) [reply ArduPlane/ArduPlane.cpp]: I got this to work, example script updated.

File: ArduPlane/ArduPlane.cpp
Comment (timtuxworth) [reply ArduPlane/ArduPlane.cpp]: Done

**Comment by timtuxworth** [comment]
I got a CI fail on sitltest-plane. I downloaded the log but can't see the problem. Can you help @IamPete1 ?

**Comment by IamPete1** [comment]
> I got a CI fail on sitltest-plane. I downloaded the log but can't see the problem. Can you help @IamPete1 ?

I have restarted it, that test has been failing occasionally.

**Comment by timtuxworth** [comment]
> > I got a CI fail on sitltest-plane. I downloaded the log but can't see the problem. Can you help @IamPete1 ?
> 
> I have restarted it, that test has been failing occasionally.

Is that something I can do myself?

**Comment by IamPete1** [comment]
> Is that something I can do myself?

Not directly, maintainers only. However, whenever you push it will restart all of CI, so you can always rebase to re-trigger everything.

**Comment by timtuxworth** [comment]
Rebased on master

**Review by tridge — APPROVED** [review]

**Pull request merged**
